services/fee-estimator/shared/utils/dynamicFees.js

```diff
@@ -13,42 +13,110 @@
  * Removal or modification of this copyright notice is prohibited.
  *
  */
-const { Logger } = require('lisk-service-framework');
+const BluebirdPromise = require('bluebird');
+
+const {
+	CacheRedis,
+	Logger,
+} = require('lisk-service-framework');
+
+const {
+	getEstimateFeeByteForBlock,
+} = require('./dynamicFeesLIP');
+const { requestConnector } = require('./request');
 
 const config = require('../../config');
 
+const executionStatus = {
+	// false: not running, true: running
+	[config.cacheKeys.cacheKeyFeeEstFull]: false,
+	[config.cacheKeys.cacheKeyFeeEstQuick]: false,
+};
+
+const cacheRedisFees = CacheRedis('fees', config.endpoints.cache);
 const logger = Logger();
 
-const calcAvgFeeByteModes = {
-	MEDIUM: 'med',
-	HIGH: 'high',
-};
+const getEstimateFeeByteForBatch = async (fromHeight, toHeight, cacheKey) => {
+	const genesisHeight = await requestConnector('getGenesisHeight');
+	const { defaultStartBlockHeight } = config.feeEstimates;
 
-const EMAcalc = (feePerByte, prevFeeEstPerByte) => {
-	const calcExpDecay = (emaBatchSize, emaDecayRate) => (
-		1 - Math.pow(1 - emaDecayRate, 1 / emaBatchSize)).toFixed(5);
+	// Check if the starting height is permitted by config or adjust acc.
+	// Use incrementation to skip the genesis block - it is not needed
+	fromHeight = Math.max(...[defaultStartBlockHeight, genesisHeight + 1, fromHeight]
+		.filter(n => !Number.isNaN(n)));
 
-	const alpha = calcExpDecay(config.feeEstimates.emaBatchSize, config.feeEstimates.emaDecayRate);
-	logger.debug(`Estimating fees with 'α' for EMA set to ${alpha}.`);
+	const cachedFeeEstimate = await cacheRedisFees.get(cacheKey);
 
-	const feeEst = {};
-	if (Object.keys(prevFeeEstPerByte).length === 0) {
-		prevFeeEstPerByte = { low: 0, med: 0, high: 0 };
-	}
+	const cachedFeeEstimateHeight = !cacheKey.includes('Quick') && cachedFeeEstimate
+		? cachedFeeEstimate.blockHeight : 0; // 0 implies does not exist
+
+	const prevFeeEstPerByte = fromHeight > cachedFeeEstimateHeight
+		? { blockHeight: fromHeight - 1, low: 0, med: 0, high: 0 } : cachedFeeEstimate;
+
+	const range = size => Array(size).fill().map((_, index) => index);
+	const feeEstPerByte = {};
+	const blockBatch = {};
+	do {
+		/* eslint-disable no-await-in-loop */
+		const idealEMABatchSize = config.feeEstimates.emaBatchSize;
+		const finalEMABatchSize = (() => {
+			const maxEMABasedOnHeight = prevFeeEstPerByte.blockHeight - genesisHeight;
+			if (idealEMABatchSize > maxEMABasedOnHeight) return maxEMABasedOnHeight + 1;
+			return idealEMABatchSize;
+		})();
+
+		blockBatch.data = await BluebirdPromise.map(
+			range(finalEMABatchSize),
+			async i => {
+				const { header, transactions } = await requestConnector(
```

Comment:
Could you please confirm if we still return `header` and `transactions` in response for `connector. getBlockByHeight` method?